// @vorld/base-service/src/tracing/tracer.ts

import { NodeTracerProvider } from '@opentelemetry/node';
import { BatchSpanProcessor, Tracer } from '@opentelemetry/tracing';
import { JaegerExporter } from '@opentelemetry/exporter-jaeger';
import { Resource } from '@opentelemetry/resources';
// import { ParentBasedSampler } from '@opentelemetry/core';
const {
    JAEGER_AGENT_HOST = 'localhost',
    JAEGER_AGENT_PORT = '6832'
} = process.env;



const tracerProvider = (serviceName: string, version: string, build: string) => {
    const tp = new NodeTracerProvider({
        // sampler: new ParentBasedSampler(),
        resource: new Resource({
            ["service.name"]: serviceName,
            ["service.version"]: version || '-',
        }),
    });

    tp.register();
    tp.addSpanProcessor(
        new BatchSpanProcessor(
            new JaegerExporter({
                host: JAEGER_AGENT_HOST,
                port: parseInt(JAEGER_AGENT_PORT, 10),
                tags: [
                    { key: 'service.name', value: serviceName },
                    { key: `service`, value: serviceName },
                    { key: `version`, value: version },
                    { key: `build`, value: build },
                ],
            }),
        ),
    );
    return tp.getTracer(serviceName, version);
};

export let tracer: Tracer = null
export const initTracer = (name: string, version: string, build: string) => {
    tracer = tracerProvider(name, version, build);
}
