// @vorld/base-service/dist/schema/schema.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateJsonSchema = exports.generateValidator = void 0;
const ajv_1 = __importDefault(require("ajv"));
const tjs = __importStar(require("typescript-json-schema"));
const fs_1 = __importDefault(require("fs"));
const ajv = createAjv();
function generateJsonSchema(modelsPath, Model) {
    const compilerOptions = {
        strictNullChecks: true,
        // to ignore node_modules
        skipLibCheck: true,
        types: [],
    };
    const settings = {
        required: true,
        noExtraProps: true,
        defaultNumberType: 'number',
    };
    // not completed
    const typesPath = fs_1.default.existsSync(`${modelsPath}/${Model.name}.d.ts`)
        ? `${modelsPath}/${Model.name}.d.ts`
        : `${modelsPath}/${Model.name}.ts`;
    const program = tjs.getProgramFromFiles([typesPath], compilerOptions);
    const generator = tjs.buildGenerator(program, settings);
    const symbols = generator.getUserSymbols();
    return generator.getSchemaForSymbols(symbols);
}
exports.generateJsonSchema = generateJsonSchema;
function generateValidator(schema) {
    const validate = ajv.compile(schema);
    return (data) => validate(data) ? [true, null] : [false, validate.errors];
}
exports.generateValidator = generateValidator;
function createAjv() {
    return new ajv_1.default({
        verbose: true,
        removeAdditional: true,
        coerceTypes: true,
        allErrors: true,
    });
}
