// mrz-parser/src/parse/td-3/td3Fields.js
const parseDocumentCode = require('../../parsers/parseDocumentCodePassport');
const parseOptional = require('../../parsers/parseOptional');
const parsePersonalNumberCheckDigit = require('../../parsers/parsePersonalNumberCheckDigit');

const {
  documentCodeTemplate,
  issuingStateTemplate,
  lastNameTemplate,
  firstNameTemplate,
  documentNumberTemplate,
  documentNumberCheckDigitTemplate,
  nationalityTemplate,
  birthDateTemplate,
  birthDateCheckDigitTemplate,
  sexTemplate,
  expirationDateTemplate,
  expirationDateCheckDigitTemplate,
  compositeCheckDigitTemplate,
} = require('../fieldTemplates');
const createFieldParser = require('../createFieldParser');

module.exports = [
  {
    ...documentCodeTemplate,
    line: 0,
    start: 0,
    end: 2,
    parser: parseDocumentCode,
  },
  {
    ...issuingStateTemplate,
    line: 0,
    start: 2,
    end: 5,
  },
  {
    ...lastNameTemplate,
    line: 0,
    start: 5,
    end: 44,
  },
  {
    ...firstNameTemplate,
    line: 0,
    start: 5,
    end: 44,
  },
  {
    ...documentNumberTemplate,
    line: 1,
    start: 0,
    end: 9,
  },
  {
    ...documentNumberCheckDigitTemplate,
    line: 1,
    start: 9,
    end: 10,
    related: [
      {
        line: 1,
        start: 0,
        end: 9,
      },
    ],
  },
  {
    ...nationalityTemplate,
    line: 1,
    start: 10,
    end: 13,
  },
  {
    ...birthDateTemplate,
    line: 1,
    start: 13,
    end: 19,
  },
  {
    ...birthDateCheckDigitTemplate,
    line: 1,
    start: 19,
    end: 20,
    related: [
      {
        line: 1,
        start: 13,
        end: 19,
      },
    ],
  },
  {
    ...sexTemplate,
    line: 1,
    start: 20,
    end: 21,
  },
  {
    ...expirationDateTemplate,
    line: 1,
    start: 21,
    end: 27,
  },
  {
    ...expirationDateCheckDigitTemplate,
    line: 1,
    start: 27,
    end: 28,
    related: [
      {
        line: 1,
        start: 21,
        end: 27,
      },
    ],
  },
  {
    label: 'Optional field 1',
    field: 'optional1',
    line: 1,
    start: 28,
    end: 42,
    parser: parseOptional,
  },
  {
    label: 'Optional field check digit',
    field: 'optionalFieldCheckDigit',
    line: 1,
    start: 42,
    end: 43,
    related: [
      {
        line: 1,
        start: 28,
        end: 42,
      },
    ],
    parser: parsePersonalNumberCheckDigit,
  },
  {
    ...compositeCheckDigitTemplate,
    line: 1,
    start: 43,
    end: 44,
    related: [
      {
        line: 1,
        start: 0,
        end: 10,
      },
      {
        line: 1,
        start: 13,
        end: 20,
      },
      {
        line: 1,
        start: 21,
        end: 43,
      },
    ],
  },
].map(createFieldParser);
