// mrz-parser/src/parse/td-1/kenyaIdCardFields.js
const parseDocumentCode = require('../../parsers/parseDocumentCodeId');
const parseDocumentNumberOptional = require('../../parsers/parseDocumentNumberOptional');

const {
  documentCodeTemplate,
  issuingStateTemplate,
  documentNumberTemplate,
  documentNumberCheckDigitTemplate,
  birthDateTemplate,
  birthDateCheckDigitTemplate,
  sexTemplate,
  issueDateTemplate,
  compositeCheckDigitTemplate,
  lastNameTemplate,
  personalNumberTemplate,
} = require('../fieldTemplates');
const createFieldParser = require('../createFieldParser');

module.exports = [
  {
    ...documentCodeTemplate,
    line: 0,
    start: 0,
    end: 2,
    parser: parseDocumentCode,
  },
  {
    ...issuingStateTemplate,
    line: 0,
    start: 2,
    end: 5,
  },
  {
    ...documentNumberTemplate,
    line: 0,
    start: 5,
    end: 14,
    related: [
      {
        line: 0,
        start: 14,
        end: 15,
      },
      {
        line: 0,
        start: 15,
        end: 30,
      },
    ],
  },
  Object.assign(documentNumberCheckDigitTemplate, {
    line: 0,
    start: 14,
    end: 15,
    related: [
      {
        line: 0,
        start: 5,
        end: 14,
      },
      {
        line: 0,
        start: 15,
        end: 30,
      },
    ],
  }),
  {
    label: 'Optional field 1',
    field: 'optional1',
    line: 0,
    start: 15,
    end: 30,
    related: [
      {
        line: 0,
        start: 5,
        end: 14,
      },
      {
        line: 0,
        start: 14,
        end: 15,
      },
    ],
    parser: parseDocumentNumberOptional,
  },
  {
    ...birthDateTemplate,
    start: 0,
    end: 6,
    line: 1,
  },
  {
    ...birthDateCheckDigitTemplate,
    line: 1,
    start: 6,
    end: 7,
    related: [
      {
        line: 1,
        start: 0,
        end: 6,
      },
    ],
  },
  {
    ...sexTemplate,
    line: 1,
    start: 7,
    end: 8,
  },
  {
    ...issueDateTemplate,
    line: 1,
    start: 8,
    end: 14,
  },
  {
    ...personalNumberTemplate,
    line: 1,
    start: 18,
    end: 26,
    parser: parseDocumentNumberOptional,
  },
  {
    ...compositeCheckDigitTemplate,
    line: 1,
    start: 29,
    end: 30,
    related: [
      {
        line: 1,
        start: 0,
        end: 29,
      },
    ],
  },
  {
    ...lastNameTemplate,
    line: 2,
    start: 0,
    end: 30,
  },
].map(createFieldParser);
