// mrz-parser/src/parse/td-1/td1-kuwait-fields.js
const { fields } = require('./td1Fields');
const { documentNumberTemplate, documentNumberCheckDigitTemplate, personalNumberTemplate } = require('../fieldTemplates');

const parseDocumentNumber = require('../../parsers/parseDocumentNumber');
const parsePersonalNumber = require('../../parsers/parsePersonalNumber');
const parseDocumentNumberCheckDigit = require('../../parsers/parseDocumentNumberCheckDigit');

const documentNumberRules = {
  ...documentNumberTemplate,
  line: 0,
  start: 5,
  end: 14,
  parser: parseDocumentNumber,
};

const parseDocumentNumberCheckDigitRules = {
  ...documentNumberCheckDigitTemplate,
  line: 0,
  start: 14,
  end: 15,
  related: [{
    line: 0,
    start: 5,
    end: 14,
  }],
  parser: parseDocumentNumberCheckDigit,
};

const parsePersonalNumberRules = {
  ...personalNumberTemplate,
  line: 0,
  start: 15,
  end: 29,
  parser: parsePersonalNumber,
};

const kuwaitFields = fields.filter((field) => (field.field !== 'compositeCheckDigit' && field.field !== 'documentNumberCheckDigit'))
  .map((field) => (field.field === 'documentNumber' ? documentNumberRules : field));

module.exports = {
  fields: [...kuwaitFields, parseDocumentNumberCheckDigitRules, parsePersonalNumberRules],
};
