// mrz-parser/src/parse/td-1/td1-ecuador-fields.js
const { fields } = require('./td1Fields');

const parseDocumentNumber = require('../../parsers/parseDocumentNumberConcatCheckDigit');
const parseSexByExtractFunc = require('../../parsers/parseSexByExtractFunc');
const parseBirthDayDateByExtractFunc = require('../../parsers/parseBirthDayDateByExtractFunc');

const { isDocNumber, isSex, isBirth } = require('../parse-helpers');

const parseBirthDayDate = parseBirthDayDateByExtractFunc((value) => {
  const LEN_YYYY_MM_DD = 8;
  const LEN_YY_MM_DD_DIGIT = 7;
  const removeAlpha = (v) => v.replace(/[A-Z]/, '');

  const date = removeAlpha(value);

  if (date.length === LEN_YYYY_MM_DD) return date.substring(2);
  if (date.length === LEN_YY_MM_DD_DIGIT) return date.substring(0, date.length - 1);

  return date;
});

const parseSex = parseSexByExtractFunc((source) => {
  const res = source.match(/[A-Z]/);
  if (res) {
    return res[0];
  }
  return source;
});

const docNumberField = fields.find(isDocNumber);
const sexField = fields.find(isSex);
const birthField = fields.find(isBirth);

const birthRule = {
  ...birthField,
  end: 8,
  parser: parseBirthDayDate,
};

const sexRule = {
  ...sexField,
  start: 6,
  end: 9,
  parser: parseSex,
};

const docNumberRule = {
  ...docNumberField,
  parser: parseDocumentNumber,
};

const ecuadorRules = {
  documentNumber: docNumberRule,
  sex: sexRule,
  birthDate: birthRule,
};

const wrapFieldEcuador = (f) => (ecuadorRules[f.field] || f);

module.exports = {
  fields: fields.map(wrapFieldEcuador),
};
