// @vorld/base-service/dist/tracing/trace.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.traceWithAttribute = exports.traceWithParams = exports.traceDecorate = exports.withSpan2 = void 0;
const api_1 = __importStar(require("@opentelemetry/api"));
var api_2 = require("@opentelemetry/api");
Object.defineProperty(exports, "SpanKind", { enumerable: true, get: function () { return api_2.SpanKind; } });
const tracer_1 = require("./tracer");
const core_1 = require("@opentelemetry/core");
function getCurrentSpanCtx(ctx) {
    return ctx.spanCtx;
}
function injectSpanCtx(ctx, sc) {
    ctx.spanCtx = sc;
}
/**
 *
 * @param {Context} ctx - context
 * @param {string} name - span name
 * @param {api.SpanOptions} options - span config
 */
const withSpan = (ctx, name = 'uncaught_span', options) => {
    const traceCtx = getCurrentSpanCtx(ctx) || api_1.default.context.active();
    const span = tracer_1.tracer.startSpan(name, options, traceCtx);
    const newCtx = ctx.clone();
    injectSpanCtx(newCtx, api_1.default.trace.setSpan(traceCtx, span));
    return [newCtx, span];
};
/**
 * @param {Context} ctx - context
 * @param {string} name - span name
 * @param {api.SpanOptions} options - span config
 */
exports.withSpan2 = (ctx, name, carrier, options) => {
    const newCtx = ctx.clone();
    const span = api_1.context.with(api_1.default.propagation.extract(api_1.context.active(), carrier), () => {
        options.startTime = core_1.hrTime();
        const sp = tracer_1.tracer.startSpan(name, options);
        sp.spanContext().isRemote = true;
        newCtx.spanCtx = sp.spanContext();
        return sp;
    });
    return [newCtx, span];
};
class SpanBuilder {
    constructor(props) {
        this.name = props === null || props === void 0 ? void 0 : props.name;
        this.attributeFromResult = props === null || props === void 0 ? void 0 : props.attributeFromResult;
    }
    decorate(f) {
        if (typeof f !== 'function') {
            throw new Error('decorated function is not provided');
        }
        const self = this;
        const name = this.name || f.name;
        if (f.constructor.name === 'Function') {
            return function _traceWrapper(origCtx, ...args) {
                if (!origCtx)
                    return f.call(this, origCtx, ...args);
                const [ctx, span] = withSpan(origCtx, name);
                try {
                    const result = f.call(this, ctx, ...args);
                    self._addAttributes(span, result);
                    return result;
                }
                catch (e) {
                    span.setAttribute('error', true);
                    span.recordException(e);
                    throw e;
                }
                finally {
                    span.end();
                }
            };
        }
        if (f.constructor.name === 'AsyncFunction') {
            return function _traceWrapper(origCtx, ...args) {
                return __awaiter(this, void 0, void 0, function* () {
                    if (!origCtx)
                        return yield f.call(this, origCtx, ...args);
                    const [ctx, span] = withSpan(origCtx, name);
                    try {
                        const result = yield f.call(this, ctx, ...args);
                        self._addAttributes(span, result);
                        return result;
                    }
                    catch (e) {
                        span.setAttribute('error', true);
                        span.recordException(e);
                        throw e;
                    }
                    finally {
                        span.end();
                    }
                });
            };
        }
        throw new Error('decorated function is not provided');
    }
    _addAttributes(span, result) {
        if (this.attributeFromResult) {
            try {
                const value = this.attributeFromResult.extract(result);
                span.setAttribute(this.attributeFromResult.name, value);
            }
            catch (e) {
                console.warn(`Could not extract tracing attribute "${this.attributeFromResult.name}" from result: ${e}`);
            }
        }
    }
}
exports.traceDecorate = (f) => new SpanBuilder().decorate(f);
exports.traceWithParams = (spanProps) => new SpanBuilder(spanProps);
exports.traceWithAttribute = (name) => ({
    // add 'getid.' to custom attributes
    fromResult: (extract = (x) => x) => new SpanBuilder({
        attributeFromResult: {
            name: `getid.${name}`,
            extract,
        }
    }),
});
