// @vorld/base-service/dist/tracing/tracer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initTracer = exports.tracer = void 0;
const node_1 = require("@opentelemetry/node");
const tracing_1 = require("@opentelemetry/tracing");
const exporter_jaeger_1 = require("@opentelemetry/exporter-jaeger");
const resources_1 = require("@opentelemetry/resources");
// import { ParentBasedSampler } from '@opentelemetry/core';
const { JAEGER_AGENT_HOST = 'localhost', JAEGER_AGENT_PORT = '6832' } = process.env;
const tracerProvider = (serviceName, version, build) => {
    const tp = new node_1.NodeTracerProvider({
        // sampler: new ParentBasedSampler(),
        resource: new resources_1.Resource({
            ["service.name"]: serviceName,
            ["service.version"]: version || '-',
        }),
    });
    tp.register();
    tp.addSpanProcessor(new tracing_1.BatchSpanProcessor(new exporter_jaeger_1.JaegerExporter({
        host: JAEGER_AGENT_HOST,
        port: parseInt(JAEGER_AGENT_PORT, 10),
        tags: [
            { key: 'service.name', value: serviceName },
            { key: `service`, value: serviceName },
            { key: `version`, value: version },
            { key: `build`, value: build },
        ],
    })));
    return tp.getTracer(serviceName, version);
};
exports.tracer = null;
exports.initTracer = (name, version, build) => {
    exports.tracer = tracerProvider(name, version, build);
};
