// @vorld/base-service/src/schema/schema.ts
import Ajv from 'ajv';
import * as tjs from "typescript-json-schema";
import fs from 'fs'
const ajv = createAjv();

type ModelConstructor<T> = new (...args: any[]) => T;

function generateJsonSchema<T extends object>(modelsPath: string, Model: ModelConstructor<T>) {
    const compilerOptions: tjs.CompilerOptions = {
        strictNullChecks: true,
        // to ignore node_modules
        skipLibCheck: true,
        types: [],
    }

    const settings: tjs.PartialArgs = {
        required: true,
        noExtraProps: true,
        defaultNumberType: 'number',
    };

    // not completed
    const typesPath = fs.existsSync(`${modelsPath}/${Model.name}.d.ts`)
        ? `${modelsPath}/${Model.name}.d.ts`
        : `${modelsPath}/${Model.name}.ts`

    const program = tjs.getProgramFromFiles([typesPath], compilerOptions);
    const generator = tjs.buildGenerator(program, settings);
    const symbols = generator.getUserSymbols();
    return generator.getSchemaForSymbols(symbols);
}

function generateValidator(schema: any): any {
    const validate = ajv.compile(schema);
    return (data: [boolean, object]) => validate(data) ? [true, null] : [false, validate.errors];
}

function createAjv() {
    return new Ajv({
        verbose: true,
        removeAdditional: true,
        coerceTypes: true,
        allErrors: true,
    });
}

export {
    generateValidator,
    generateJsonSchema,
    ModelConstructor
};
