// mrz-parser/src/parse/formats.js
const formats = {
  TD1: 'TD1',
  TD2: 'TD2',
  TD3: 'TD3',

  BOTSWANA_TD1: 'BOTSWANA_TD1',
  DOMINICAN_TD1: 'DOMINICAN_TD1',
  ECUADOR_TD1: 'ECUADOR_TD1',
  GUATEMALA_TD1: 'GUATEMALA_TD1',
  KUWAIT_TD1: 'KUWAIT_TD1',
  MOLDOVA_TD1: 'MOLDOVA_TD1',
  MOZAMBIQUE_TD1: 'MOZAMBIQUE_TD1',
  PAKISTAN_DOMESTIC_TD1: 'PAKISTAN_DOMESTIC_TD1',
  POLAND_TD1: 'POLAND_TD1',
  SALVADOR_TD1: 'SALVADOR_TD1',
  URUGUAY_TD1: 'URUGUAY_TD1',

  SWISS_DRIVING_LICENSE: 'SWISS_DRIVING_LICENSE',
  FRENCH_NATIONAL_ID: 'FRENCH_NATIONAL_ID',
  FRENCH_DRIVING_LICENSE: 'FRENCH_DRIVING_LICENSE',
  ESTONIAN_DRIVING_LICENSE: 'ESTONIAN_DRIVING_LICENSE',
  IRELAND_DRIVING_LICENSE: 'IRELAND_DRIVING_LICENSE',
  NETHERLANDS_DRIVING_LICENSE: 'NETHERLANDS_DRIVING_LICENSE',
  SLOVAK_DRIVING_LICENSE: 'SLOVAK_DRIVING_LICENSE',
  CROATIA_DRIVING_LICENSE: 'CROATIA_DRIVING_LICENSE',
  MEXICAN_ID_CARD: 'MEXICAN_ID_CARD',
  CHILE_NATIONAL_ID: 'CHILE_NATIONAL_ID',
  KENYA_ID_CARD: 'KENYA_ID_CARD',
  LUXEMBOURG_ID_CARD: 'LUXEMBOURG_ID_CARD',
};
Object.freeze(formats);

module.exports = formats;
