// mrz-parser/src/parse/parse.js
const { ValidationError } = require('./exceptions');

const { validateMrzInput } = require('./checkLines');
const parsers = require('./parsers');

const getParser = (line, parsersList) => (
  parsersList.find(({ pattern }) => line.match(pattern)));

const parsersByFormat = {
  TD1: [
    { pattern: /^[ACI].BWA/, parser: parsers.BOTSWANA_TD1 },
    { pattern: /^IDCHL/, parser: parsers.CHILE_NATIONAL_ID },
    { pattern: /^IDDOM/, parser: parsers.DOMINICAN_TD1 },
    { pattern: /^IDECU/, parser: parsers.ECUADOR_TD1 },
    { pattern: /^IDKYA/, parser: parsers.KENYA_ID_CARD },
    { pattern: /^I.GTM/, parser: parsers.GUATEMALA_TD1 },
    { pattern: /^I.KWT\d*[A-Z]+/, parser: parsers.KUWAIT_TD1 },
    { pattern: /^IDMEX/, parser: parsers.MEXICAN_ID_CARD },
    { pattern: /^CAMDA/, parser: parsers.MOLDOVA_TD1 },
    { pattern: /^[A-Z<]{2}MOZ/, parser: parsers.MOZAMBIQUE_TD1 },
    { pattern: /^I.PAK\d*[A-Z]+/, parser: parsers.PAKISTAN_DOMESTIC_TD1 },
    { pattern: /^I.POL/, parser: parsers.POLAND_TD1 },
    { pattern: /^IDSLV/, parser: parsers.SALVADOR_TD1 },
    { pattern: /^I.URY/, parser: parsers.URUGUAY_TD1 },

    { pattern: /./, parser: parsers.TD1 },
  ],
  TD2: [
    { pattern: /^I.FRA/, parser: parsers.FRENCH_NATIONAL_ID },
    { pattern: /./, parser: parsers.TD2 },
  ],
  TD3: [
    { pattern: /./, parser: parsers.TD3 },
  ],
  DRIVER_LICENSE: [
    { pattern: /^D[1PN<]FRA/, parser: parsers.FRENCH_DRIVING_LICENSE },
    { pattern: /^D[1PN<]EST/, parser: parsers.ESTONIAN_DRIVING_LICENSE },
    { pattern: /^D[1PN<]NLD/, parser: parsers.NETHERLANDS_DRIVING_LICENSE },
    { pattern: /^DLSVK/, parser: parsers.SLOVAK_DRIVING_LICENSE },
    { pattern: /^D[1PN<][A-Z<]{12}[A-Z]{3}/, parser: parsers.IRELAND_DRIVING_LICENSE },
    { pattern: /^D[1PN<][0-9<]{27}[0-9]$/, parser: parsers.CROATIA_DRIVING_LICENSE },
    { pattern: /^[A-Z]{3}\d{3}[A-Z<]{3}$/, parser: parsers.SWISS_DRIVING_LICENSE },
  ],
  LUXEMBOURG_ID_CARD: [
    { pattern: /^[\d<.]{38}$/, parser: parsers.LUXEMBOURG_ID_CARD },
  ],
};

const defineFormat = (firstLine) => {
  if (firstLine.match(/^D[1PN<L][0-9A-Z<]{27}[0-9]$|^[A-Z]{3}\d{3}[A-Z<]{3}$/)) {
    return 'DRIVER_LICENSE';
  }
  switch (firstLine.length) {
    case 30: return 'TD1';
    case 36: return 'TD2';
    case 38: return 'LUXEMBOURG_ID_CARD';
    case 44: return 'TD3';
    default: return null;
  }
};

function parseMRZ(mrzLines) {
  const lines = validateMrzInput(mrzLines);
  if (lines.length > 3) {
    throw new ValidationError('Unknown document format. Mrz row count must be less than 3');
  }

  const format = defineFormat(lines[0]);
  if (!format) {
    throw new ValidationError('unrecognized document format. First line of input must have 30 (TD1), 36 (TD2 or French National Id), 44 (TD3) or 9 (Swiss Driving License) characters');
  }

  const { parser } = getParser(lines[0], parsersByFormat[format]) || {};
  if (!parser) {
    throw new ValidationError('unrecognized document format.');
  }
  return parser(lines);
}

module.exports = parseMRZ;
