// mrz-parser/src/ocr-template/ocr-categories.js
module.exports = {
  firstName: {
    category: 'First name',
    order: 1,
    contentType: 'string',
    crosschecking: true,
    compare: {
      startsWith: true,
      intersect: true,
    },
    required: true,
    aliases: [
      'First Name',
    ],
  },
  lastName: {
    category: 'Last name',
    order: 2,
    contentType: 'string',
    crosschecking: true,
    compare: {
      startsWith: true,
      intersect: true,
    },
    required: true,
    aliases: [
      'Last Name',
      'Surname',
    ],
  },
  documentType: {
    category: 'Document type',
    order: 5,
    contentType: 'string',
    enum: [
      'passport',
      'id-card',
      'driving-licence',
      'residence-permit',
      'voter-card',
    ],
    crosschecking: true,
    aliases: [
      'Doc Type',
    ],
  },
  nationality: {
    category: 'Nationality',
    contentType: 'string',
    crosschecking: true,
    aliases: [
      'Doc Type',
    ],
  },
  nationalityCode: {
    category: 'Nationality code',
    contentType: 'country',
    order: 10,
    crosschecking: true,
  },
  gender: {
    category: 'Gender',
    order: 3,
    contentType: 'sex',
    crosschecking: true,
    enum: [
      'male',
      'female',
    ],
    aliases: [
      'Sex',
    ],
  },
  documentNumber: {
    category: 'Document number',
    order: 6,
    contentType: 'string',
    mrzChecksum: 'documentNumberCheckDigit',
    crosschecking: true,
    aliases: [
      'Doc Number',
    ],
  },
  country: {
    category: 'Issue country',
    order: 7,
    contentType: 'country',
    crosschecking: true,
    aliases: [
      'Country',
    ],
  },
  dateOfBirth: {
    category: 'Date of birth',
    order: 4,
    contentType: 'date',
    mrzChecksum: 'birthDateCheckDigit',
    crosschecking: true,
    aliases: [
      'Date of Birth',
      'Date Of Birth',
    ],
  },
  dateOfIssue: {
    category: 'Date of issue',
    order: 8,
    contentType: 'date',
    crosschecking: true,
    aliases: [
      'Date of Issue',
      'Date Of Issue',
    ],
  },
  dateOfExpiry: {
    category: 'Date of expiry',
    order: 9,
    contentType: 'date',
    mrzChecksum: 'expirationDateCheckDigit',
    crosschecking: true,
    required: true,
    aliases: [
      'Expiry',
      'Date Of Expiry',
      'Date of Expiry',
    ],
  },
  placeOfBirth: {
    category: 'Place of birth',
    order: 11,
    contentType: 'string',
    crosschecking: true,
    aliases: [
      'Place of Birth',
      'Place Of Birth',
    ],
  },
  placeOfIssue: {
    category: 'Place of issue',
    order: 12,
    contentType: 'string',
    crosschecking: true,
    aliases: [
      'Issuing State Name',
    ],
  },
  mrzFull: {
    category: 'MRZ full',
    order: 13,
    contentType: 'string',
    editable: true,
    aliases: [
      'MRZ Full',
    ],
  },
  personalNumber: {
    category: 'Personal number',
    order: 14,
    contentType: 'string',
    mrzChecksum: 'compositeCheckDigit',
    aliases: [
      'Personal Number',
    ],
  },
  optionalData: {
    category: 'Optional data',
    order: 16,
    contentType: 'string',
  },
  optionalData2: {
    category: 'Optional data (part 2)',
    order: 17,
    contentType: 'string',
  },
};
