// @vorld/base-service/dist/context.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, privateMap, value) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to set private field on non-instance");
    }
    privateMap.set(receiver, value);
    return value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, privateMap) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to get private field on non-instance");
    }
    return privateMap.get(receiver);
};
var _values;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Context = void 0;
class Context {
    constructor() {
        _values.set(this, void 0);
        this.setValue = (dataKey, dataValue) => {
            __classPrivateFieldGet(this, _values)[dataKey] = dataValue;
        };
        this.getValue = (dataKey) => {
            return __classPrivateFieldGet(this, _values)[dataKey];
        };
        __classPrivateFieldSet(this, _values, {});
    }
    clone() {
        const ctx = new Context();
        ctx.spanCtx = this.spanCtx;
        __classPrivateFieldSet(ctx, _values, Object.assign({}, __classPrivateFieldGet(this, _values)));
        return ctx;
    }
    debug(message) {
        console.debug(`DEBUG: ${message}`);
    }
    info(message) {
        console.debug(message);
    }
    error(message) {
        console.debug(message);
    }
}
exports.Context = Context;
_values = new WeakMap();
