// mrz-parser/src/controllers/index.js
const { ValidationError } = require('../parse/exceptions');
const parse = require('../parse/parse');
const { normalizeMrz } = require('../parse/normalize-by-ocr');
const categories = require('../ocr-template/ocr-categories');

async function parseMrz(ctx, { mrz }) {
  try {
    return parse(mrz);
  } catch (error) {
    if (error instanceof ValidationError) {
      ctx.info(error.message);
      return {
        valid: false,
        validationError: error.message,
      };
    }
    ctx.error(error.message);
    return { valid: false };
  }
}

async function normalizeMrzByOcr(ctx, { mrz, ocr = [] }) {
  const { mrzFull } = categories;
  try {
    return await normalizeMrz(mrz, ocr);
  } catch (error) {
    if (error instanceof ValidationError) {
      ctx.info(error.message);
      return [
        {
          category: mrzFull.category,
          content: mrz,
          contentType: mrzFull.contentType,
          valid: false,
        },
      ];
    }
    ctx.error(error.message);
    return [
      {
        category: mrzFull.category,
        content: mrz,
        contentType: mrzFull.contentType,
        valid: false,
      },
    ];
  }
}

module.exports = {
  parseMrz, normalizeMrzByOcr,
};
